src/benchq/compilation/ruby_slippers.jl

```diff
@@ -0,0 +1,649 @@
+state generated by a circuit using a graph state simulator (graph_sim) from the paper
+"Fast simulation of stabilizer circuits using a graph state representation" by Simon
+Anders, Hans J. Briegel. https://arxiv.org/abs/quant-ph/0504117". We have modified
+the algorithm to ignore paulis.
+
+Many of the modules have a no_teleport version. This is to denote that the compiler
+does not use teleportation in these cases and is more or less just it's predicessor,
+a graph simulator for stabilizer states.
+=#
+
+using PythonCall
+using StatsBase
+
+include("graph_sim_data.jl")
+
+const AdjList = Set{Int32}
+
+const Qubit = UInt32
+
+struct ICMOp
+    code::LCO
+    qubit1::Qubit
+    qubit2::Qubit
+
+    ICMOp(name, qubit) = new(name, qubit, 0)
+    ICMOp(name, qubit1, qubit2) = new(name, qubit1, qubit2)
+end
+
+struct RubySlippersHyperparams
+    teleportation_threshold::UInt16
+    teleportation_distance::UInt8
+    min_neighbors::UInt8
+    max_num_neighbors_to_search::UInt32
+    decomposition_strategy::UInt8
+end
+
+default_hyperparams = RubySlippersHyperparams(40, 4, 6, 1e5, 1)
+
+
+"""
+Converts a given circuit in Clifford + T form to icm form and simulates the icm
+circuit using the graph sim mini simulator. Returns the adjacency list of the graph
+state created by the icm circuit along with the single qubit operations on each vertex.
+teleportation_threshold, min_neighbors, teleportation_distance, and  max_num_neighbors_to_search
+are metaparameters which can be optimized to speed up the simulation.
+
+Args:
+    circuit::Circuit  circuit to be simulated
+
+Returns:
+    adj::Vector{AdjList}              adjacency list describing the graph state
+    lco::Vector{LCO}                  local clifford operations on each node
+    teleportation_threshold::Int      max node degree allowed before state is teleported
+    min_neighbors::Int                stop searching for neighbor with low degree if
+                                        neighbor has at least this many neighbors
+    teleportation_distance::Int        number of teleportations to do when state is teleported
+    max_num_neighbors_to_search::Int  max number of neighbors to search through when finding
+                                        a neighbor with low degree
+"""
+function run_ruby_slippers(
+    circuit,
+    verbose=false,
+    max_graph_size=0,
+    teleportation_threshold=40,
+    teleportation_distance=4,
+    min_neighbors=6,
+    max_num_neighbors_to_search=1e5,
+    decomposition_strategy=1,
+)
+    max_graph_size = pyconvert(UInt32, max_graph_size)
+
+    # params which can be optimized to speed up computation
+    hyperparams = RubySlippersHyperparams(
+        teleportation_threshold,
+        teleportation_distance,
+        min_neighbors,
+        max_num_neighbors_to_search,
+        decomposition_strategy,
+    )
+
+    if max_graph_size == 0
+        max_graph_size = get_max_n_nodes(circuit, hyperparams.teleportation_distance)
+    end
```

Comment:
I think it might be counterintuitive that setting `max_graph_size` to zero results in a special value being used for the max graph size. Would it make sense to use `max_graph_size=nothing` or alternatively make a separate method that doesn't have the `max_graph_size` parameter?